src/lib/plot.typ

```diff
@@ -19,95 +20,71 @@
-  }
-
-  let lin-interpolated-pt(a, b) = {
-    let x1 = a.at(0)
-    let y1 = a.at(1)
-    let x2 = b.at(0)
-    let y2 = b.at(1)
-
-    /* Special case for vertical lines */
-    if x2 - x1 == 0 {
-      return (x2, calc.min(max-y, calc.max(y2, min-y)))
-    }
-
-    if y2 - y1 == 0 {
-      return (calc.min(max-x, calc.max(x2, min-x)), y2)
-    }
-
-    let m = (y2 - y1) / (x2 - x1)
-    let n = y2 - m * x2
-
-    let x = x2
-    let y = y2
-
-    y = calc.min(max-y, calc.max(y, min-y))
-    x = (y - n) / m
-
-    x = calc.min(max-x, calc.max(x, min-x))
-    y = m * x + n
-
-    return (x, y)
+/// Sample a function with one parameter
+///
+/// - fn (function): Function to sample
+/// - domain (array): X domain tuple (min, max)
+/// - samples (int): Number of samples
+/// - sample-at (array): Sample function at values
+/// -> array: Array of (x, y) points
+#let sample-fn(fn, domain, samples, sample-at: ()) = {
+  assert(samples >= 0,
+    message: "Samples must be positive or zero")
+  assert(samples + sample-at.len() >= 2,
+    message: "You must at least sample 2 values")
+  assert(type(domain) == array and domain.len() == 2,
+    message: "Domain must be a tuple")
+
+  let (lo, hi) = domain
+
+  let y0 = (fn)(lo)
+  let is-vector = type(y0) == array
+  if not is-vector {
+    y0 = ((lo, y0), )
+  } else {
+    y0 = (y0, )
   }
 
-  let paths = ()
+  let pts = sample-at + range(0, samples + 1).map(t => lo + t / samples * (hi - lo))
+  pts = pts.sorted()
 
-  let path = ()
-  let prev-p = none
-  for p in data {
-    if p == none { continue }
-
-    let (px, py, ..) = p
-    if px == none or py == none { continue }
-
-    if in-range(p) {
-      if not in-range(prev-p) and prev-p != none {
-        path.push(lin-interpolated-pt(p, prev-p))
-      }
-
-      path.push(p)
+  return pts.map(x => {
+    if is-vector {
+      (fn)(x)
     } else {
-      if in-range(prev-p) {
-        path.push(lin-interpolated-pt(prev-p, p))
-      } else if prev-p != none {
-        let a = lin-interpolated-pt(p, prev-p)
-        let b = lin-interpolated-pt(prev-p, p)
-        if in-range(a) and in-range(b) {
-          path.push(a)
-          path.push(b)
-        }
-      }
-
-      if path.len() > 0 {
-        paths.push(path)
-        path = ()
-      }
+      (x, (fn)(x))
     }
+  })
+}
 
-    prev-p = p
-  }
-
-  if path.len() > 0 {
-    paths.push(path)
-  }
-  return paths
+/// Sample a function with two parameters
+///
+/// - fn (function): Function to sample
+/// - x-domain (array): X domain tuple (min, max)
+/// - y-domain (array): X domain tuple (min, max)
+/// - x-samples (int): Number of samples
+/// - y-samples (int): Number of samples
+/// - x-sample-at (array): Sample function at x values
+/// - y-sample-at (array): Sample function at y values
+/// -> array: Array of z values
+#let sample-fn2(fn, x-domain, y-domain, x-samples, y-samples) = {
+  assert(x-samples >= 2 and x-samples <= 500)
+  assert(y-samples >= 2 and y-samples <= 500)
```

Comment:
This is not really true because `sample-fn` allows passing manual sample points, so passing a sample-count of 0 is fine. I removed the max. limit of samples for both functions.